src/RECONR/ReferenceGrid/src/call.hpp

```diff
@@ -1,65 +1,172 @@
-std::vector< double > operator()( const RP::SpecialCase& sc ) const {
-  return { sc.EL(), sc.EH() };
+std::array<double, 3>
+operator()( const resolved::BreitWigner::Lvalue::Resonance& resonance ) const {
+  const double energy = resonance.ER();
+  const double halfWidth = 0.5 * resonance.GT();
+  return {{ energy - halfWidth, energy, energy + halfWidth }};
 }
 
-std::array<double, 3> operator()( 
-  const RP::resolved::BreitWigner::Lvalue::Resonance& resonance ) const {
+std::array<double, 3>
+operator()( const resolved::ReichMoore::Lvalue::Resonance& resonance ) const {
+  const double energy = resonance.ER();
+  const double halfWidth = 0.5 * ( resonance.GN()
+                                   + resonance.GG()
+                                   + resonance.GFA()
+                                   + resonance.GFB() );
+  return {{ energy - halfWidth, energy, energy + halfWidth }};
+}
 
-  const double E = resonance.ER();
-  const double hw = resonance.GT()*0.5;
-  return {{E-hw, E, E+hw}};
+template< typename ResolvedLvalue,
+          if_< hasResonances< ResolvedLvalue > > = true >
+auto
+operator()( const ResolvedLvalue& lValue ) const {
+  return lValue.resonances() | ranges::view::transform( *this );
 }
 
-auto operator()( 
-  const RP::resolved::ReichMoore::Lvalue::Resonance& resonance ) const {
+template< typename Range, if_< hasLvalues< Range > > = true >
+auto operator()( const Range& range ) const {
+  std::vector< double > energies;
+  energies.reserve( 12 * range.lValues().size() );
+
+  const auto lowerLimit = range.EL();
+  const auto upperLimit = nudgeDown( range.EH() );
+
+  energies.push_back( lowerLimit );
+
+  {
+    auto resonances = range.lValues() | ranges::view::for_each( *this );
+    RANGES_FOR( auto resonance, resonances ){
+      ranges::action::push_back( energies,
+                                 resonance
+                                 | ranges::view::filter
+                                   ( [=]( const auto energy )
+                                     { return lowerLimit < energy
+                                              and energy < upperLimit; } ) );
+    }
+  }
 
-  const double E = resonance.ER();
-  const double hw = 0.5*(resonance.GN() + 
-                         resonance.GG() +
-                         resonance.GFA() +
-                         resonance.GFB() );
-  buffer = {{E-hw, E, E+hw}};
-  return ranges::make_iterator_range( buffer.begin(), buffer.end() );
+  energies.push_back( upperLimit );
+
+  if ( not ranges::is_sorted( energies ) ){
+    orlp::pdqsort( energies.begin(), energies.end() );
+  }
+
+  energies |= ranges::action::unique;
```

Comment:
I'm not sure what this syntax does. Does it just ensure that all the values are unique? What does the `|=` syntax doe?

Reply:
you can think of this as short hand for
```cpp
energies = std::move(energies) | range::action::unique;
``
which is short for
```cpp
energies = range::action::unique( std::move(energies) );
``
which is short for

```cpp
energies.erase( std::unique( energies.begin(), energies.end() ), energies.end() );
```

which takes an sorted list and removes duplicate entries.